nsq/async.py

```diff
@@ -277,7 +277,7 @@ def _read_body(self, data):
             self.trigger(event.DATA, conn=self, data=data)
         except Exception:
             logger.exception('uncaught exception in data event')
-        self.io_loop.add_callback(self._start_read)
```

Comment:
I'm pretty sure it won't "block", it'll either immediately handle unprocessed (but read) bytes or it will defer and setup a callback.

http://www.tornadoweb.org/en/stable/_modules/tornado/iostream.html#BaseIOStream.read_bytes

This just skips an _additional_ deferment.